src/main/java/com/google/sps/servlets/MealServlet.java

```diff
@@ -177,23 +177,45 @@ private void getMealById(HttpServletRequest request, HttpServletResponse respons
     }
 
     private void returnIdOfSimilar(HttpServletRequest request, HttpServletResponse response) throws IOException {
-        // Fuction redirect to random page
+        // Fuction redirect to random page of the same type
         // TODO(grenlayk): implement suggestions algorithm here for Product Alpha
         Long mealId = Long.parseLong(getParameter(request, "id", "0"));
+        String mealType = null;
 
         Query query = new Query("Meal");
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
         PreparedQuery results = datastore.prepare(query);
+
+        for (Entity entity : results.asIterable()) {
+            Long id = (Long)entity.getProperty("id");
+            if (id != null && id.equals(mealId)) {
+                mealType = (String)entity.getProperty("type");
+            }
+        }
```

Comment:
Instead of requesting all data and going through it one by one, please `addFilter` to the Query that will filter out by id being equal to `mealId`. You can look up `getMealById` for how to do that.